cpp/src/arrow/record_batch.h

```diff
@@ -234,6 +234,53 @@ class ARROW_EXPORT RecordBatchReader {
     return batch;
   }
 
+  class RecordBatchReaderIterator {
+   public:
+    RecordBatchReaderIterator() : batch_(RecordBatchEnd()) {}
+
+    explicit RecordBatchReaderIterator(RecordBatchReader* reader)
+        : batch_(RecordBatchEnd()), reader_(reader) {
+      Next();
+    }
+
+    bool operator!=(const RecordBatchReaderIterator& other) const {
+      return batch_ != other.batch_;
+    }
+
+    Result<std::shared_ptr<RecordBatch>> operator*() {
+      ARROW_RETURN_NOT_OK(batch_.status());
+
+      auto batch = std::move(batch_);
+      batch_ = RecordBatchEnd();
```

Comment:
It is rather weird for `operator*` to change the stored batch. That means if I use this operator twice, I will get two different results; that sounds rather undesired.